File: src/0147_insertion_sort_list.js
function ListNode(val = 0, next = null) {
    this.val = val;
    this.next = next;
}

function insertionSortList(head) {
    if (!head || !head.next) return head;
    const sentinel = new ListNode();
    let curr = head;
    while (curr) {
        let prev = sentinel;
        while (prev.next && curr.val >= prev.next.val) prev = prev.next;
        const nextCurr = curr.next;
        curr.next = prev.next;
        prev.next = curr;
        curr = nextCurr;
    }
    return sentinel.next;
}

module.exports = { insertionSortList };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(listNodeToString(insertionSortList(new ListNode(4, new ListNode(2, new ListNode(1, new ListNode(3, null)))))), listNodeToString(new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, null))))));
    assert.deepStrictEqual(listNodeToString(insertionSortList(new ListNode((-1), new ListNode(5, new ListNode(3, new ListNode(4, new ListNode(0, null))))))), listNodeToString(new ListNode((-1), new ListNode(0, new ListNode(3, new ListNode(4, new ListNode(5, null)))))));
}

File: src/0147_insertion_sort_list.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { val, next: None }
    }
}

struct Solution;

impl Solution {
    pub fn insertion_sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let values = list_to_vec(&head);
        let mut sorted: Vec<i32> = Vec::new();
        for value in values {
            let mut index = 0;
            while index < sorted.len() && value >= sorted[index] {
                index += 1;
            }
            sorted.insert(index, value);
        }
        vec_to_list(sorted)
    }
}

fn list_to_vec(head: &Option<Box<ListNode>>) -> Vec<i32> {
    let mut values = Vec::new();
    let mut current = head.as_ref();
    while let Some(node) = current {
        values.push(node.val);
        current = node.next.as_ref();
    }
    values
}

fn vec_to_list(values: Vec<i32>) -> Option<Box<ListNode>> {
    let mut head = None;
    for value in values.into_iter().rev() {
        let mut node = Box::new(ListNode::new(value));
        node.next = head;
        head = Some(node);
    }
    head
}

#[cfg(test)]
mod tests {
    use super::*;

    fn list(values: &[i32]) -> Option<Box<ListNode>> {
        values
            .iter()
            .rev()
            .fold(None, |next, &val| Some(Box::new(ListNode { val, next })))
    }

    fn list_values(head: &Option<Box<ListNode>>) -> Vec<i32> {
        let mut values = Vec::new();
        let mut current = head.as_deref();
        while let Some(node) = current {
            values.push(node.val);
            current = node.next.as_deref();
        }
        values
    }

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            list_values(&Solution::insertion_sort_list(list(&[4, 2, 1, 3]))),
            vec![1, 2, 3, 4]
        );
        assert_eq!(
            list_values(&Solution::insertion_sort_list(list(&[-1, 5, 3, 4, 0]))),
            vec![-1, 0, 3, 4, 5]
        );
    }
}
